Porting odML GUI from Python 2 to Python 3

digaru19 opened the pull request:
The following changes are for porting the odML GUI editor from Python 2 to Python 3, switching over to PyGObject library. The editor will be based on odML library version **1.3**. It will be compatible with both Python 2 and Python 3.

These are some of the major/minor things that are to be accomplished along the process, for completing the task. This list will be updated 'extensively' along the process.  

**TODOs** :- 
- [x] Separate GUI codebase from core library codebase
- [X] Fix _"some"_ common Python 2 to Python 3 porting issues
- [X] Start the porting process using _pygtkcompat_
- [X] Disable certain functionalities temporarily, which are independent of the core GUI functions
- [X] Fix _MenuItem_ stock and labels
- [x] Fix _menu item_ click events
- [x] Fix new document creation method through the _Wizard_ 
- [x] Fix _TreeModel_ components in Python 2 for the GUI editor
- [ ] Fix tab-switching 
- [ ] Enable _Drag-and-Drop_ providers
- [ ] Enable _Recently Used Files_ suggestions
  (Much more to come) 

P.S. :- The commit history will be cleaned after the whole process completes :sweat_smile:  .

digaru19 committed:
Add .gitignore file

digaru19 committed:
Separate GUI code from core-library code

digaru19 committed:
Import pygtkcompat for gtk 3+

digaru19 committed:
Fix common Py2 to Py3 porting issues

Fix 'print' usages;
Fix module imports;
Fix urllib imports;
Fix hashlib imports and algorithms;

digaru19 committed:
Make general changes for porting; Disable some functionalities temporarily;

Change data-access methods
Disable Drag-and-Drop support [Temp]
Disable recently used files suggestions [Temp]
Change file opening method in Gio [Temp]

digaru19 committed:
Fix Model<-->odML resolutions

digaru19 committed:
Fix right-click context menu

Fix the popup menu on right-click in SectionView and PropertyView.
Change 'stock' usage to represent 'stock-id'

digaru19 committed:
Circumvent stock items registration; Fix menu item labels;

The stock item registration seems to fail.
Hence, in places where stock-id could give a
complete description of the menu item, we have
to pass the parameters explicitly.

achilleas-k commented:
Haven't read through the whole thing extensively but I'd like to note something regarding the change in the file structure. The `gui/` directory moving out of `odml/` into the root of the repository is the first step towards separating the module, but it would be good to rename it now, when it's still early. In the end, this will be the module root, so perhaps it should be renamed to `odmlui` or something similar. Since it'll end up being the actual name of the module/package, perhaps we could decide on it now before we end up with a name no one likes.

digaru19 commented:
Sure. :+1: 
What should be the name then, `odmlui` ?

jgrewe commented:
I'm fine with odmlui,
just noticed that there are some *.pyc bin files under version control ;)

achilleas-k commented:
odmlui it is, then!

On Mon, 12 Jun 2017, 18:43 Jan Grewe, <notifications@github.com> wrote:

> I'm fine with odmlui,
> just noticed that there are some piyc bin files under version control ;)
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/G-Node/odml-ui/pull/4#issuecomment-307846709>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACQmrVl6vNxjBGXdrqBBnLc69qIQGNJ1ks5sDWqxgaJpZM4N3Kdr>
> .
>

digaru19 committed:
Change module name from 'gui' to 'odmlui'

digaru19 committed:
Fix odML terminology loading methods

Fix hashing method for terminology cache file
Fix import methods for xmlparser

digaru19 committed:
Fix 'File-->Open' action to open files

digaru19 committed:
Fix right-click terminology popup menu

digaru19 committed:
Fix document saving method

digaru19 committed:
Fix odML terminology cache saving

digaru19 committed:
Fix document rendering in GUI for Python 2

digaru19 committed:
Fix tab-switching method

Change reparenting procedure, as previous method is deprecated
Temporarily disable state-restore method (need to fix TreePath Conversions)

digaru19 committed:
Correct variable names; Add comments;

digaru19 commented on odmlui/DocumentRegistry.py:
```diff
@@ -19,7 +20,8 @@ def add(self, doc):
         """
-        id = hash(doc)
+        # odML.Document is unhashable. Converting to string before hashing
+        id = hash(str(doc))
```
Or maybe we should change it to `hash( id(doc) )` ?

achilleas-k replied:
if Document is unhashable and we need it to be, it should implement `__hash__()`, right?

digaru19 replied:
Yeah, to make it hashable 'internally', we need to have a `__hash__()` method.
But if we do implement one, what do we put in it ? I mean, how should be the `hash` computed ? If we would again consider the author name, version, etc. in the internal hash function, then I thought that maybe just convert the `doc` into string (which already has a `__hash()__` method), and that would have most of the above mentioned information.
So, do we implement a separate hash function ?

digaru19 committed:
Enable 'Recently Used Files' feature

achilleas-k commented on odmlui/treemodel/GenericIter.py:
```diff
@@ -41,7 +41,7 @@ def get_next(self):
         or None
         """
-        obj = next(self._obj)
+        obj = self._obj.__next__()
```
Was this changes to accommodate Python 2? I know `next()` is available in 2.6+. Are we supporting 2.5?
Calling `__next__()` is kinda ugly.

digaru19 replied:
The line being deleted here (red one) was introduced by me, in the starting commits of this PR. :sweat_smile: 
Initially, it was `self._obj.next()` which works fine in Python 2, but has issues in Python 3. 
Python 3 requires the `next(self._obj)` method. So, the `next()` method worked in Py 3, but not in Py 2. Hence, in this commit, I changed it to the most specific function name. 
In Python 2, to make `next()` work, it seems that we need to implement a `next()` method in the object, and not `__next__()`. So, to just have the standard `next()` method and be Py2/3 compatible, we need to alias `need` to `__need__` method in the `nodes.py` in odML core library. So, instead of making a change there, I made the change here. 
If you feel that the `next()` method is more suitable, do let me know. I will make the change in the `python-odml` repo, and use `next()` method here.

achilleas-k commented:
I've been through the current state and made a few comments. Since we're merging into a development branch, I'd vote for going ahead and merging (if @jgrewe, @mpsonntag, and @gicmo agree) and making issues for the unchecked items in the first comment as well as any other comments we have.
